File: 0152-maximum-product-subarray/0152-maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int res = nums[0];
        int currMin = 1;
        int currMax = 1;

        for (int num : nums) {
            int temp = currMax * num;

            currMax = Math.max(currMax * num, Math.max(num, currMin * num));
            currMin = Math.min(temp, Math.min(num, currMin * num));
            res = Math.max(currMax, res);
        }

        return res;
    }
}

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        n = len(nums)

        max_dp = nums[0]
        min_dp = nums[0]
        res = nums[0]
        for i in range(1, n):
            num = nums[i]

            if num < 0:
                max_dp, min_dp = min_dp, max_dp

            max_dp = max(num, num*max_dp)
            min_dp = min(num, num*min_dp)

            res = max(res, max_dp)

        return res